ucp/_libs/ucx_api.pyx

```diff
@@ -446,11 +446,88 @@ cdef class UCXListener(UCXObject):
         return int(<uintptr_t>self._handle)
 
 
+# Counter used as UCXRequest UIDs
+cdef int _ucx_request_counter = 0
+
+cdef class UCXRequest:
+    """Python wrapper of UCX request handle.
+
+    Notice, this class doesn't own the handle and multiple instances of
+    UCXRequest can point to the same underlying UCX handle.
+    Furthermore, UCX can modify/free the UCX handle without notice
+    thus we use `_uid` to make sure the handle hasn't been modified.
+    """
+    cdef:
+        ucp_request *_handle
+        int _uid
+
+    def __init__(self, uintptr_t req_as_int):
+        global _ucx_request_counter
+        cdef ucp_request *req = <ucp_request*>req_as_int
+        assert req != NULL
+        self._handle = req
+
+        cdef dict info = {"status": "pending"}
+        if self._handle.info == NULL:  # First time we are wrapping this UCX request
+            Py_INCREF(info)
+            self._handle.info = <PyObject*> info
+            _ucx_request_counter += 1
+            self._uid = _ucx_request_counter
+            assert self._handle.uid == 0
+            self._handle.uid = _ucx_request_counter
+        else:
+            self._uid = self._handle.uid
+
+    def closed(self):
+        return self._handle == NULL or self._uid != self._handle.uid
+
+    def close(self):
+        """This routine releases the non-blocking request back to UCX,
+        regardless of its current state. Communications operations associated with
+        this request will make progress internally, however no further notifications or
+        callbacks will be invoked for this request. """
+
+        if not self.closed():
+            Py_DECREF(<object>self._handle.info)
+            self._handle.info = NULL
+            self._handle.uid = 0
+            ucp_request_free(self._handle)
+            self._handle = NULL
+
+    @property
+    def info(self):
+        assert not self.closed()
+        return <dict> self._handle.info
+
+    @property
+    def handle(self):
+        assert not self.closed()
+        return int(<uintptr_t>self._handle)
+
+    def __hash__(self):
+        if self.closed():
+            assert False
+            return id(self)
```

Comment:
Do we need to `return` if we already `assert False`? Also if the intent is to error when this closed, could we add a more descriptive error (perhaps a `ValueError`)?

Reply:
Sorry, the `assert False` is leftovers from debugging :)

Reply:
No worries. Thanks for clarifying/fixing! :)